compiler: remove support for the legacy GCs

zerbina opened the pull request:
## Summary

There are multiple issues with the legacy memory management strategies
(the ones that are not destructor-based):

* each one introduces another language dialect
* they necessitate the old runtime (more code to maintain)
* they're specific to the C backend, making them not fit in well with
  the direction of having each backend support the full language
* the way they're currently integrated with the compiler introduces a
  large amount of complexity and conditional logic to the C code-
  generator

For the reasons listed above, compiler support for both the legacy GCs
and large parts of the old runtime is removed. The only remaining
selectable GC modes are `orc` and `arc`.

Except for the v1 run-time-type-information, which is still required by
the `deepcopy` implementation and `marshal` module, the old runtime is
now obsolete and can be removed.

## Details

**CLI:**

* adjust the allowed `gc` switch arguments to only include
  `destructors`, `arc`, and `orc`
* remove the obsolete `--seqsv2` option (the version 2 implementation is
  always used now, at least for the C backend)

**C code-generator:**

* remove all usages of both `optSeqDestructors` and `optTinyRtti` --
  they're now unconditionally enabled when using the C backend
* remove `optTinyRtti`, as it has no remaining usage-sites
* remove all code related to the old runtime's `string`s and `seq`s
* remove all support for the garbage collectors using the old runtime
* remove code-generation logic for the unused magics `mIntToStr`,
  `mInt64ToStr`, and `mFloatToStr`
* simplify the code generated for the entry point

**Tests:**

* remove tests specific to the removed GCs
* disable the tests for `--gc:none`
* don't specify `--gc:arc` for some tests where not required, allowing
  the tests to be joined

**Documentation:**

* clean up the memory management documentation (`gc.rst`)
* remove some mentions of the removed GCs and the old runtime

**CI:**

* remove `boehm`-related package dependencies

**Misc:**

* remove `when` statements related to the removed GCs from compiler code
* support `koch` building the first iteration with `--gc:orc`, which is
  a preparation for updating the csources compiler to one that doesn't
  support `--gc:refc`

For the purpose of backwards compatibility, the now removed GC modes can
still be tested for with `compileOption`.

### Removal of `--gc:none`

It is currently not clear how `--gc:none` should work under the new
runtime, as no traditional GC that can be disabled by simply not running
it exists. Sequences and strings are also no longer ref-based, making it
unclear how these should work.

For the time being, the option is removed, but with the intention to
bring it back once the semantics are decided.

zerbina committed:
cgen: remove `optSeqDestructors` usage

The option indicates whether version 2 sequences are enabled, which is
always the case for the C backend after the legacy GCs are removed.

Note that the flag was misused for checking whether a destructor-based
GC is enabled, meaning that some of the removals here are more related
to the GC removal than to the seq v1 removal.

zerbina committed:
compiler: remove support for the legacy GCs

This includes the options, enums, and the implementation in the code
generator.

zerbina committed:
cgen: remove `optTinyRtti` and usages of it

The option is implied when using `--gc:arc|orc` and is thus obsolete
now, as it's never queried outside of the C code-generator.

zerbina committed:
cgen: remove unused magics

Neither the compiler nor `system` have made use of the magics for some
time now, so it would make sense to remove their remanining usages in
a follow-up PR.

zerbina committed:
compiler: remove mentions of the legacy GCs

Also remove usages of defines and option queries related to them.

zerbina committed:
cgen: remove code generation for the v1 `repr`

It's part of the old runtime.

zerbina committed:
remove now unused reports

zerbina committed:
remove mentions and usage of `-d:useRealtimeGC`

It's a feature only available with `refc`.

zerbina committed:
tests: fix the `tcstring.nim` test

It is now tested with `--gc:orc` (previously only with `--gc:arc`),
which triggers code to be run that previously wasn't. The
necessary `prepareMutation` calls are inserted in order to make the test
work.

zerbina committed:
tests: adjust the test suite

* remove usages of the legacy destructors from test specifications
* remove tests only relevant to the legacy GCs
* don't use `--gc:arc` where not required for the test

zerbina committed:
docs: update various documentation

Remove mentions of the old runtime and legacy GCs. It's likely that
there's still now-stale documentation remaining somewhere.

zerbina committed:
ci: don't install boehm-related packages for testing

They're not needed anymore.

zerbina committed:
ccgstmts: remove dead code

The code was already unused prior to the removal of the legacy GCs.

zerbina committed:
koch: fix the ORC bootstrapping test

The first iteration unconditionally used `--gc:refc`, which is a problem
when the bootstrapping compiler is one that no longer supports
`--gc:refc`.

The newer compiler versions (which properly support `--gc:orc`,
although not necessarily default to it) identify themselves as "Nimskull
Compiler" (compared to "Nim Compiler"), and `koch` now uses this
information in order to decide whether `--gc:refc` needs to be
enabled.

zerbina committed:
update tests and documentation after `--gc:none` removal

zerbina committed:
cli: remove the `--seqsv2` option

It's purpose is to only enable the destructor-based `seq`s and
`string`s, but nothing else. The destructor-based one is now the default
(and only) available implementation for the C backend, and since both
the JavaScript and VM backend don't support destructor-based sequences
yet and would fail to compile if forced to use it, the option is
removed.

saem commented on compiler/backend/ccgliterals.nim:
```diff
@@ -93,9 +60,4 @@ proc genNilStringLiteral(m: BModule; info: TLineInfo): Rope =
-  else:
-    internalError(
-      m.config, n.info, "cannot determine how to produce code for string literal")
+  result = genStringLiteralV2(m, n, isConst = true)
```
Removing all the `V2` suffix stuff is part of the runtime removal?

zerbina replied:
Yep, my thinking was that it's cleaner to remove all the suffixes (both in the compiler and runtime) at once.

saem replied:
That does sound cleaner.

saem commented on compiler/backend/ccgstmts.nim:
```diff
@@ -15,21 +15,6 @@ const
   stringCaseThreshold = 8
     # above X strings a hash-switch for strings is generated
 
-proc getTraverseProc(p: BProc, v: PSym): Rope =
```
It's fantastic how much code we're dropping here, further drives the idea that our backends should greatly simplify and we move more of the brains earlier in compilation.

saem commented on compiler/front/commands.nim:
```diff
@@ -61,6 +46,14 @@ from compiler/ast/report_enums import ReportKind,
+  # Helper to build boot constants, for debugging you can 'echo' the else part.
+  const name = if expr: " " & userString else: ""
+
+bootSwitch(usedRelease, defined(release), "-d:release")
```
Being very pedantic the imports can't see these, so timing wise it's happening a bit late. But I don't think there are any real dependencies on them right now. Might be a latent bug.

zerbina replied:
I'm not sure if that's what you mean, but the imports also weren't able to see these previously, as the constants are not exported.

saem replied:
Sorry, I got my wires crossed, I barely slept yesterday. Anyhow, what I meant was that the other modules could play around with the defines before we get a chance to capture them here.

zerbina replied:
Hm, that's indeed possible, though I think such behaviour (i.e., defines inside a module persisting) should be disallowed.

I'll move them back up and leave a note.

zerbina replied:
Hm, that's indeed possible, though I think such behaviour (i.e., defines inside a module persisting) should be disallowed.

I'll move them back up and leave a note.

saem commented on tests/ccgbugs/tgc_bug.nim:
```diff
@@ -1,5 +1,4 @@
 discard """
-matrix: "--gc:refc -d:fulldebug -d:smokeCycles"
```
Why drop the defines while keeping them in `twrongrefcounting.nim`?

zerbina replied:
I'm not entirely certain, but believe that it's an oversight -- I'm going to restore the defines here.

saem commented on tests/lang_callable/iter/titer2.nim:
```diff
@@ -8,9 +8,13 @@ discard """
 3
 4'''
-  cmd: "nim $target --gc:none --hints:on --warnings:off $options $file"
+  #cmd: "nim $target --gc:none --hints:on --warnings:off $options $file"
```
```suggestion
  #matrix: "--gc:none --hints:on --warnings:off"
```

Just to phasing out more of these `cmd`s; doesn't have to be done figured while the line is being edited regardlesss.

saem reviewed (approved):
This is great!

A really nice benefit of this is that we can start standardizing the gdb debug script all that much more easily and include it in our CI testing. That should give us decent debugging and making exploring/learning so much easier.

alaviss commented:
So with this I persume that we are gonna build the compiler with ORC by default?

If so then we should remove the `orc` job from CI configuration.

zerbina commented:
> So with this I persume that we are gonna build the compiler with ORC by default?

How do you mean? The compiler is already built with `--gc:orc`, see #547. The reason I didn't remove the CI job was to keep ensuring that all three bootstrap iterations work with ORC, as the current csources compiler is not able to build the compiler with `--gc:orc` and thus still uses `--gc:refc`.

> If so then we should remove the `orc` job from CI configuration.

Yep, will do.

zerbina committed:
tests: small test spec adjustments

saem commented:
Did we want to cut a sources before or after this change?

zerbina commented:
> Did we want to cut a sources before or after this change?

If the aim is full coverage at all times, then it's probably better to update the csources *before* removing the ORC bootstrap job. I'm indifferent, and okay with restoring the `orc` job  again.

saem commented:
IIRC, the csources are generated every time we merge into `devel` as one of the artifacts. So if the job removal is not part of this PR, then we'll maintain coverage, can use this PR's very own release for csources. Meaning a follow-up change can drop the job and update the csources as an atomic operation.

zerbina committed:
commands: partially revert `bootSwitch` change

If one of the imported modules alters the defines in some way (they
really shouldn't, but it's not impossible), it would affect the
`bootSwitch`es.

Processing the boot switches is moved to above the imports again.

saem commented:
I'm pretty sure this is good to go, feel free to merge if you're all set, @zerbina.

zerbina commented:
/merge

All set. It's time to merge.
